docs/src/nondiff_points.md

```diff
@@ -0,0 +1,149 @@
+# What to return for non-differentiable points
+!!! info "What is the short version?"
+    If the function is not differentiable due to e.g. a branch, like `abs`, your rule can reasonably claim the derivative at that point is the value from either branch, *or* any value in-between (e.g. for `abs` claiming 0 is a good idea).
+    If it is not differentiable due to the primal not being defined on one side, you can set it to what ever you like.
+    Your rule should claim a derivative that is *useful*.
+In calculus one learns that if the derivative as computed by approaching from the left,
+and the derivative one computes as approaching from the right are not equal then the derivative is not defined,
+and we say the function is not differentiable at that point.
+This is distinct from the notion captured by [`NoTangent`](@ref), which is that the tangent space itself is not defined: because in some sense the primal value can not be perturbed e.g. is is a discrete type.
+
+However, contrary to what calculus says most autodiff systems will return an answer for such functions.
+For example for: `abs_left(x) = (x <= 0) ? -x : x`, AD will say the derivative at `x=0` is `-1`.
+Alternatively for:  `abs_right(x) = (x < 0) ? -x : x`, AD will say the derivative at `x=0` is `1`.
+Those two examples are weird since they are equal at all points, but AD claims different derivatives at `x=0`.
+The way to fix autodiff systems being weird is to write custom rules.
+So what rule should we write for this case?
+
+The obvious answer, would be to write a rule that throws an error if input at a point where calculus says the derivative is not defined.
+Another option is to return some error signally value like `NaN`.
+Which you *can* do.
+However, this is not useful.
+Instead we introduce what we call the **sub/super-differential convention**:
+
+> It is always permissible to return any element of the sub/super-differential.
+> You should return the most useful.
+
+Or equivalently but considering the trivial singleton sub/super-differential seperately:
+
+> At any point where the derivative is not defined, it is permissible to return any element of the sub/super-differential.
+> You should return the most useful.
+
+We will justify this further below, but first let us discuss what it means.
+
+## What is the sub/super-differential?
+
+The subderivative is defined only for locally convex functions, where-as the super-deriviative is fined only for locally concave functions.
+For our purpose we, basically never care which we are working with  and so write sub/super-derivative.
+
+For a function $f$ at some point $x_0$ a sub/super-derivative is a real number $c$ such that there exists a open ball $\mathcal{B} \subset \mathrm{dom}(f)$ containing $x_0$,
```

Comment:
it definately should be zero.
Derivative from left is zero, derivative from right is zero
I will look closer at the math later and see why it is not including it

Reply:
yeah, what I meant to write is that maybe the formula above only applies to continuous functions

Reply:
yeah, I think subdifferentials are not defines at discontinualities.
Because being concave or convex requires continuous.

Reply:
yeah, I think subdifferentials are not defines at discontinualities.
Because being concave or convex requires continuous.

Reply:
Actually, the definition of convexity/concavity doesn't care about continuity. 

The pragmatic thing is to extend the function to be lower semicontinuous, which means letting `f(0) = 0`. This also gives you local convexity, so that everything in this doc makes sense.